pymongo/lock.py

```diff
@@ -0,0 +1,126 @@
+    # - Atomic increment lock counter. (done first and last to ensure
+    # correct state)
+    # - Acquire lock.
+    # - Critical region (and any nested critical regions)
+    # - Release lock.
+    # - Decrement lock counter.
+
+    if platform.python_implementation() == "CPython":
+        _locks: weakref.WeakSet = weakref.WeakSet()  # References to instances of _ForkLock
+        _global_lock = threading.Lock()  # Used for indicating we're about to fork.
+        _global_locked_cv = threading.Condition()  # Used for notifying locks.
+
+        # Atomic counter for locked locks.
+        _atomic_counter = 0
+        _atomic_counter_lock = threading.Lock()
+
+        _insertion_lock = threading.Lock()
+
+        def __init__(self):
+            self._lock = threading.Lock()
+            with _ForkLock._insertion_lock:
+                _ForkLock._locks.add(self)
+
+        def __getattr__(self, item):
+            return getattr(self._lock, item)
+
+        def __enter__(self):
+            # Decide whether we're about to fork.
+            if _ForkLock._atomic_counter > 0:  # Possible dependent lock, keep going
+                with _ForkLock._atomic_counter_lock:  # Increment first.
+                    _ForkLock._atomic_counter += 1
+                self._lock.__enter__()
+            else:  # Possible we might fork, there's no other locks.
+                with _ForkLock._global_locked_cv:
+                    while _ForkLock._global_lock.locked():  # About to fork?
+                        _ForkLock._global_locked_cv.wait()  # Wait until we finish.
+                        # Parent => Will continue as before
+                        # Child => Caller will die anyways.
+                    with _ForkLock._atomic_counter_lock:  # Increment first.
+                        _ForkLock._atomic_counter += 1
+                    self._lock.__enter__()
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            self._lock.__exit__(exc_type, exc_val, exc_tb)
+            with _ForkLock._atomic_counter_lock:  # Make sure the counter is updated
+                _ForkLock._atomic_counter -= 1
+
+        @classmethod
+        def _release_locks(cls):
+            # Completed the fork, reset all the locks.
+            for lock in cls._locks:
+                lock._lock.release()
+            _ForkLock._global_lock.release()  # Release our lock
+            with _ForkLock._global_locked_cv:
+                _ForkLock._global_locked_cv.notify_all()
+                # Notify all waiting threads in the parent.
+                # In the child, there should be no waiting threads as only the
+                # calling thread gets forked.
+
+        @classmethod
+        def _acquire_locks(cls):
+            _ForkLock._global_lock.acquire()  # Indicate we're about to fork.
+            while _ForkLock._atomic_counter > 0:
+                pass  # Spin while the other threads finish up.
+            for lock in cls._locks:  # Acquire them all.
```

Comment:
I don't think we should try to acquire _all_ the locks since that is vulnerable to deadlock. Let's only acquire the locks protecting the WeakSets.